cpp/include/cudf/ast/expressions.hpp

```diff
@@ -552,6 +554,98 @@ class column_name_reference : public expression {
   std::string column_name;
 };
 
+/**
+ * @brief An AST expression tree. it owns and contains multiple dependent expressions. All the
+ * expressions are destroyed once the tree is destructed.
+ */
+class tree {
+ public:
+  /**
+   * @brief construct an empty ast tree
+   */
+  tree() = default;
+
+  /**
+   * @brief Moves the ast tree
+   */
+  tree(tree&&) = default;
+
+  /**
+   * @brief move-assigns the AST tree
+   * @returns a reference to the move-assigned tree
+   */
+  tree& operator=(tree&&) = default;
+
+  ~tree() = default;
+
+  // the tree is not copyable
+  tree(tree const&)            = delete;
+  tree& operator=(tree const&) = delete;
+
+  /**
+   * @brief Add an expression to the AST tree
+   * @param args Arguments to use to construct the ast expression
+   * @returns a reference to the added expression
+   */
+  template <typename Expr, typename... Args>
+  Expr const& emplace(Args&&... args)
+  {
+    static_assert(std::is_base_of_v<expression, Expr>);
+    auto expr            = std::make_shared<Expr>(std::forward<Args>(args)...);
+    Expr const& expr_ref = *expr;
+    expressions.emplace_back(std::static_pointer_cast<expression>(std::move(expr)));
+    return expr_ref;
+  }
+
+  /**
+   * @brief Add an expression to the AST tree
+   * @param expr AST expression to be added
+   * @returns a reference to the added expression
+   */
+  template <typename Expr>
+  Expr const& push(Expr expr)
+  {
+    return emplace<Expr>(std::move(expr));
+  }
+
+  /**
+   * @brief get the first expression in the tree
+   * @returns the first inserted expression into the tree
+   */
+  expression const& front() const { return *expressions.front(); }
+
+  /**
+   * @brief get the last expression in the tree
+   * @returns the last inserted expression into the tree
+   */
+  expression const& back() const { return *expressions.back(); }
+
+  /**
+   * @brief get the number of expressions added to the tree
+   * @returns the number of expressions added to the tree
+   */
+  size_t size() const { return expressions.size(); }
+
+  /**
+   * @brief get the expression at an index in the tree. index is checked.
+   * @param index index of expression in the ast tree
+   * @returns the expression at the specified index
+   */
+  expression const& at(size_t index) { return *expressions.at(index); }
+
+  /**
+   * @brief get the expression at an index in the tree. index is unchecked.
+   * @param index index of expression in the ast tree
+   * @returns the expression at the specified index
+   */
+  expression const& operator[](size_t index) const { return *expressions[index]; }
+
+ private:
+  // TODO: use better ownership semantics, the shared_ptr here is redundant. consider using a bump
+  // allocator with type-erased deleters.
+  std::vector<std::shared_ptr<expression>> expressions;
```

Comment:
Are we storing a shared_ptr here in anticipation of using a bump allocator and storing non-owning pointers/references to the values in that allocator? With the current design I guess the main reason we need to store a vector of shared_ptr is because the vector could be reallocated and get moved so we can't just make this a `std::vector<expression>` and return stable references to elements of the vector? If that's the main reason, what about using a `std::list<expression>` instead?

Reply:
no. we are using `std::shared_ptr` here so we can store and type-erase the actual expression type via `std::static_pointer_cast` i.e . We can't store a `std::list<expression>` as `expression` is a virtual base class and `std::list` stores the concrete type. 
We don't need shared ownership here at all since the ast tree owns the expressions. But if we allocate memory pools for storing the objects, it would be better and more efficient, as the reference counting is redundant.
And if we need shared ownership, we can always wrap the expression tree with a `shared_ptr`, which is what we always want as the expressions can't live without one another. 
`std::vector` is only used here for sequential access or index-based referencing of the expressions.
if we use a bump/object pool allocator, we'd only need the bump allocator and `std::vector<expression *>`, the memory will then be cleaned up on destruction of the ast and the type-erased expressions destroyed.

Reply:
Ahh right sorry I forgot about the inheritance structure. OK yes that makes sense then.